### DISCUSSION TITLE
Inline styles as attributes versus in stylesheet - Has anyone tried using `juice`?

### DISCUSSION THREAD

**Discussion by universemaster** [KaTeX/KaTeX#3367] [General]
[Juice](https://www.npmjs.com/package/juice) is a tool:

>Given HTML, juice will inline your CSS properties into the style attribute.

Given that I read once somewhere on [web.dev/](https://web.dev/) that matching selectors takes 50% of the style-related processing, I wondered if inlining would help.  (It would obviously increase file size, too, though).

I failed to use Juice because I think it baulked at my 20MB file.

Has anyone experimented in this area?

**Comment by edemaine** [comment]
Juice looks cool &mdash; I've wanted to be able to embed KaTeX in email, and that's a partial solution. Unfortunately, I don't think fonts can be properly embedded in email, so it might look ugly.

I haven't tried it to increase rendering performance, because I generally don't have an issue with rendering performance. But seems worth a try... Maybe try with a 1MB subset of your file and see how it compares?

**Comment by universemaster** [comment]
> I don't think fonts can be properly embedded in email, so it might look ugly.

Yes, I've also noticed had issues with fonts with [SingleFile](https://github.com/gildas-lormeau/SingleFile), which I believe is used by [Zotero](https://www.zotero.org/).

And, similarly, the download webpage button on [chrome mobile](https://support.google.com/chrome/answer/7343019?hl=en-GB&co=GENIE.Platform%3DAndroid) seems make the page look bad with for some of the fonts.

I'm keen to enable offline use (and I have no desire to try and page-split it properly into a PDF - see chrome bug related to [css grid fragmentation](https://bugs.chromium.org/p/chromium/issues/detail?id=614667#c33), similarly for `display:flex` actually, and for the bad support for [css-page-break](https://caniuse.com/css-page-break).
